### PR TITLE
계산기 [STEP 1] redmango

### PR DESCRIPTION
**Pull request by redmango1447**
@JoSH0318 안녕하세요 밈
잠은 오고 Unit Test에 관한 생각은 이것저것 드는데 선뜻 키보드 위로 손이 안가 우선 pr을 보내보기로 했습니다 ㅎㅎ...

# STEP 1
### `Queue`와 `Linked List`
다양한 `Queue`와 `List` 자료구조 형태가 있지만 무엇을 선택하여 사용해야 할지 고민했습니다. 우선 현재 프로젝트 구현에 필요한 기능은 차례대로 하나씩 넣고 뽑아주는 기능이라고 생각했습니다. 요구 조건에 연산 우선순위를 지키지 않는다고 했고 그렇다면 탐색 및 중간에서 값을 뽑아 쓰는 기능은 불필요하다고 판단한 것입니다. 그렇다면 결국 데이터의 삽입 및 삭제가 빠르다고 하는 `Linked List`를 사용해야 하는가 하고 고민해 봤지만 아래와 같은 이유로 더블 스택을 이용한 `Queue`로 구현해 봤습니다.
1. 단순 추가 삭제 기능으로 한정했을 경우 `Linked List`와 더블스택 큐 두 방법에서 유의미한 성능 차이가 있다는 증거를 찾을 수 없었습니다.
2. `Linked List`의 경우 더블스택 큐에 비해 상대적으로 구현해야 할 코드가 많고 복잡합니다. 모든 경우가 그런 건 아니겠지만 제가 생각했을 때 일반적으로 구현해야 할 코드가 많고 복잡하다는 것은 그만큼 에러가 발생할 경우의 수가 높아지고 유지보수에 문제가 생긴다고 생각합니다. 현재의 경우는 필요한 기능이 단순해서 발생하지 않을 문제일 수 있지만 그럼에도 상대적으로 구현 규모와 복잡도가 낮은 더블스택 큐가 좀 더 적합하다고 생각되었습니다.

### 제네릭
구현을 하고 보니 다양한 타입을 받아 사용할 수 있는 제네릭을 사용할지 고민했습니다.
하지만 `View`에서 받아올 값들을 `String`타입으로 하나의 `Queue`에 일괄적으로 다 넣어야 하는지 아니면 `Int`,`Double`,`Sring` 타입별로 `Queue`를 생성해서 넣어야 하는지, 즉 계산 기능 구현을 어떻게 해야 하는지 감을 못 잡았기에 우선은 구현시 사용한 String 타입을 유지하기로 하고 추후에 필요하면 변경하기로 결정했습니다.

### `TDD` 적용시 커밋 단위
지금까진 TDD에 맞게
1. 실패할 테스트 구현
2. 코드 리팩토링
3. 테스트 성공

위와 같은 순서로 구현은 했지만 커밋은 테스트 성공 시에만 했습니다.
하지만 `TDD`의 굳이 순서가 있다는 걸 생각해 본다면 다른 분들처럼 실패할 테스트를 구현하고 1번, 테스트 성공 시 1번 이렇게 하는 게 맞을까요?

### `Unit Test` 세분화 및 이미 테스트한 메소드 사용
세분화에 대한 것은 사실 끝도 없을 것 같지만 어느 정도 선이란 게 있을까 조언을 구하고 싶습니다. 또한 테스트에 이전 테스트에 통과한 메서드를 사용해도 되는 건지 알고 싶습니다

### CODE REVIEW

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        // then
+        XCTAssertFalse(expectedValue)
+    }
+
+    func test_dequeue_enqueueStack의_첫번째로_들어간_값을_반환한다() {
+        // given
+        let input = ["1", "2", "3"]
+        for item in input {
+            sut.enqueue(item)
+        }
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: 반복문도 좋다고 생각합니다.
test는 주어진 상황이 단순하고, 명확히 보여야한다고 생각합니다.
따라서 enqueue를 3번 호출하는 것이 좀 더 좋아보여요!
이부분에 대해서 어떻게 생각하시나요?

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        // given
+
+        //when
+        let expectedValue: String? = nil
+        
+        // then
+        XCTAssertEqual(sut.dequeue(), expectedValue)
+    }
+    
+    func test_clear_모든_stack의_값을_제거한다() {
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: 메서드명이 이전 test 메서드들과 다르게 명확하지 않아요. 
`제거한다`라는 행위의 예상값은 무엇인가요?

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    }
+    
+    func test_dequeue_stack의_값이_없을때_dequeue를_하면_nil이_반환된다() {
+        // given
+
+        //when
+        let expectedValue: String? = nil
+        
+        // then
+        XCTAssertEqual(sut.dequeue(), expectedValue)
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: XCTAssertNil을 사용하지 않고 XCTAssertEqual을 사용한 이유가 있을까요?

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        _ = sut.dequeue()
+        sut.enqueue("1")
+        
+        sut.clear()
+        
+        //when
+        let expectedValue = true
+        
+        // then
+        XCTAssertEqual(sut.isEmpty, expectedValue)
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: 위 test 메서드에선 XCTAssertFalse를 사용하고있는데,
XCTAssertTrue를 사용하지 않고 XCTAssertEqual을 사용한 이유가 있을까요?

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+final class CalculatorTests: XCTestCase {
+    var sut: CalculatorItemQueue!
+
+    override func setUpWithError() throws {
+        sut = CalculatorItemQueue()
+    }
+
+    override func tearDownWithError() throws {
+        sut = nil
+    }
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: super를 호출하지 않았습니다! 특별히 이유가 있을까요?
super를 호출한다는 것은 어떠한 의미인가요?

**Review by JoSH0318 — COMMENTED** [review]
안녕하세요 😀 레드망고🥭
첫번째 pr하느라 고생하셨습니다.
자료구조에 대해서 고민하신 부분에서 많은 어려움이 있었겠구나 생각이 들어요.
또한, test 관련 코드도 아직 익숙하지 못한 것일 뿐이니 너무 걱정하지 않으셔도 된다 생각해요.
지금도 충분히 잘 짜여진 코드라고 생각합니다! 조금만 더 익숙해지고 다듬어보면 좋을 것 같습니다!

몇가지 코멘트, 질문, 요구사항, 답변 남겼습니다!
작성하신 코드에 대한 충분한 근거가 있다면 모든 사항을 수정하실 필요없습니다. 다만, 그 근거로 저를 설득시켜주세요!👍

---

### 질문
1️⃣ Queue와 Linked List
고민한 점 > Queue와 Linked List에서 `단순 추가 삭제 기능으로 한정했을 경우 Linked List와 더블스택 큐 두 방법에서 유의미한 성능 차이가 있다는 증거를 찾을 수 없었습니다.`라고 말씀하셨습니다. 
그렇다면 Queue와 Linked List의 시간복잡도, 공간복잡도, 특징, 장단점은 무엇인가요? 이점을 찾아보신다면 유의미한 근거를 찾을 수 있다고 생각합니다. 

2️⃣ 파일트리, Protocol
현재 Protocol을 따로 그룹으로 파일 분리를 해주셨는데, 이렇게 정리해주신 이유가 궁금해요!
만약 Controller 영역 관련 protocol이 생긴다면 이 protocol은 어디에 위치하게 될가요?🤔

---

### 요구사항
1️⃣ 제네릭
말씀하신대로 오직 String만 넣어준다면 문제될 게 없습니다. 하지만 `오직` Strung만 취급한다는게 단점이 될 것 같아요. 확장성과 범용성이 떨어진다고 생각합니다. 
물론 해당 프로젝트에서는 redmango의 의견대로 String만을 취급해도 문제될 게 없을 수도 있겠네요! 
하지만 개인적으로 제네릭을 사용해서 확장성과 범용성을 고려한 더 나은 코드를 지향하는게 좋다고 생각해요.
redmango의 생각은 어떤가요?

2️⃣ Unit Test
test를 좀 더 세분화하고 다양한 경우와 edge case를 고려한 test가 이뤄지면 좋을 것 같아요.😀 아직 익숙하지 못해서 다양한 경우가 떠오르지 않는 것 뿐이라고 생각해요! 조금만 더 생각하면 좋을 것 같습니다!

3️⃣ UML
작성해주신 UML이 익숙하지 못한 디자인인 것 같아요! 

---

### 답변
1️⃣ TDD 적용시 커밋 단위
Q: 지금까진 TDD에 맞게 `실패할 테스트 구현`>`코드 리팩토링`>`테스트 성공`
위와 같은 순서로 구현은 했지만 커밋은 테스트 성공 시에만 했습니다.
하지만 TDD의 굳이 순서가 있다는 걸 생각해 본다면 다른 분들처럼 실패할 테스트를 구현하고 1번, 테스트 성공 시 1번 이렇게 하는 게 맞을까요?

> A: 사실 이부분은 개인과 팀의 주관에 따라 달라질 것 같아요. 개인적인 견해로는 `어쨋든 두가지 방법 모두가 TDD를 적용했다.`가 중요하다고 생각합니다. TDD기반으로 코드를 작성해서 TDD의 이점을 취했으면 그만이라 생각합니다. 따라서 어떠한 방법이든 상관 없다고 생각해요. 
지난번 redmango의 의견은 `Red`(실패를 위한 test)는 오류가 발생하고, 오류가 있는 코드를 commit하는 건 옳지 않다 라는 생각이셨던 것 같습니다. 맞나요? 😀 그렇다면 `Red Green` 묶어서 commit하고 `Refactor`를 따로 commit하겠네요! 저는 redmango의 생각 또한 옳다고 생각합니다.
하지만 누군가는 `Red Test 또한 기능 추가의 시작이고 commit을 남겨야한다, test 코드는 오류가 나도 상관없다` 고 생각하시는 분도 있을 것 같네요. 그렇다면 commit 단위를 나눠야할 것 같네요!
뭐든 본인만의 근거가 있다면 다 좋다고 생각합니다!

2️⃣ Unit Test 세분화 및 이미 테스트한 메소드 사용
Q: 세분화에 대한 것은 사실 끝도 없을 것 같지만 어느 정도 선이란 게 있을까 조언을 구하고 싶습니다. 또한 테스트에 이전 테스트에 통과한 메서드를 사용해도 되는 건지 알고 싶습니다.

> A: 일단 `어느 정도 선`이라는 것을 말씀드리기 어려운 부분인 것 같아요! test는 정말 여러가지 상황이 있을테니 상황마다 `선`을 설정하기 어려우니까요🥹. 하지만 개인적인 견해로는 `충분히 해당 기능을 test하고 있는지`가 중요할 것 같아요. 그러기 위해서는 여러 상황을 가정하고 test해보는게 좋을 것 같아요. 예를 들어 `isEmpty`를 test하기 위해서 고려해봐야할 상황은 1) enqueue, dequeue 과정이 없는 초기 상황 2) enqueue, dequeue 과정을 거친 상황을 고려해봐야하겠네요!
이쯤해서 그 다음 질문인 enqueue, dequeue는 아직 test를 하지 않은 상황인 test에 사용해도 되는 것인가에 대한 의문이 생길 것 같습니다. 이 부분은 enqueue, dequeue test를 통과한 후에, refactor 과정에서 추가해주면 된다고 생각합니다.

File: Calculator/CalculatorTests/CalculatorTests.swift (hunk 1 above)
Comment (redmango1447) [reply Calculator/CalculatorTests/CalculatorTests.swift]: 지금과 같은 호출이 적은 상황에서는 밈의 말이 맞는것 같습니다! 다만 혹시 UnitTest에서 이런 경우가 있는진 모르겠지만 만약  enqueue를 1만번 호출해야한다면 그땐 아래와 같이 하는게 더 명확하게 보일까요?
```swift
for number in 1...10000 {
    sut.enqueue(number)
}
```

File: Calculator/CalculatorTests/CalculatorTests.swift (hunk 2 above)
Comment (redmango1447) [reply Calculator/CalculatorTests/CalculatorTests.swift]: isEmpty는 true, count로 보면 0이 되겠군요. 그렇다면 test_clear_모든_stack의_값을_제거한다 -> `test_clear_clear를_호출하면_isEmpty프로퍼티는_true값을_가진다` 는 어떨까요?

File: Calculator/CalculatorTests/CalculatorTests.swift (hunk 3 above)
Comment (redmango1447) [reply Calculator/CalculatorTests/CalculatorTests.swift]: 이것은 순전히 UnitTest와 XCT메소드에 익숙하지 않아 생긴 휴먼에러?입니다 ㅠㅠ

File: Calculator/CalculatorTests/CalculatorTests.swift (hunk 4 above)
Comment (redmango1447) [reply Calculator/CalculatorTests/CalculatorTests.swift]: 이것도 위 답변과 같이 UnitTest와 XCT메소드에 익숙하지 않아 생긴 휴먼에러?입니다 헣ㅎ....다음부턴 어떤 메소드가 있는지 파악하여 상황에 맞게 잘 사용해보겠습니다!

File: Calculator/CalculatorTests/CalculatorTests.swift (hunk 5 above)
Comment (redmango1447) [reply Calculator/CalculatorTests/CalculatorTests.swift]: super는 자신의 슈퍼클래스, 다시말해 부모 클래스를 가리킵니다. super를 호출한다는것은 부모클래스의 원래 기능을 사용하겠다는 의미인걸로 알고있습니다. 호출하지 않은 특별한 이유는 없습니다만 반대로 호출해야하는 특별한 이유가 있었다면 TestTarget생성시 기본 템플릿으로 제공해주지 않았을까 싶습니다.(예를들어 ViewController파일 생성시 ViewDidLoad()메소드 내부에 super.viewDidLoad()가 이미 호출된것처럼 말이죠)
또한 공식문서는 아니지만 이번에 관련된 정보를 찾는도중 도움이 될만한 정보를 찾았습니다.
https://stackoverflow.com/questions/53448538/xcode-10-and-super-teardown?noredirect=1&lq=1
제가 잘 이해한게 맞는지는 모르겠지만 저런 특수 케이스를 제외하면 문제가 없기 때문에 템플릿 생성시 기본적으로 제공하지 않는걸까 생각이 듭니다. 하지만 글에서도 나와있듯 특수 케이스는 드물지만 있는일이고 그때 문제가 될 여지가 있으니 앞으로 UnitTest시에 관련 메소드들에 super를 호출하는 습관을 들이는게 맞는 것 같습니다.

**Comment by redmango1447** [comment]
> ### 질문
> 1️⃣ Queue와 Linked List 고민한 점 > Queue와 Linked List에서 `단순 추가 삭제 기능으로 한정했을 경우 Linked List와 더블스택 큐 두 방법에서 유의미한 성능 차이가 있다는 증거를 찾을 수 없었습니다.`라고 말씀하셨습니다. 그렇다면 Queue와 Linked List의 시간복잡도, 공간복잡도, 특징, 장단점은 무엇인가요? 이점을 찾아보신다면 유의미한 근거를 찾을 수 있다고 생각합니다.
> 

약간의 고민과 검색으로 아래와 같은 표을 만들어 봤습니다!

| | DoubleStack Queue | Single LinkedList Queue|
| :--------: | :--------: | :--------: |
| 특징     | enqueueStack(입력받을 스택)과 dequeueStack(입력받은 스택을  뒤집어서 담은 스택)으로 이루어짐     | 연속적이지 않은 공간에 데이터와 다음 노드의 위치 정보를 담은 Node와 현재 위치의 Node를 가리키는 head로 이루어짐   |
| 장점     | 단순 추가 삭제 기능만을 가진다면 구현이 쉽고 직관적이다. 접근,탐색이 LinkedList에 비해 상대적으로 용이하다   | 추가 삭제가 자유롭다.    |
| 단점     | 2개의 스택을 사용함으로써 메모리 비용이 크고 중간에 값을 추가하거나 삭제할 경우 overhead가 발생한다.<br>또한 reverced해서 dequeueStack을 생성하는 과정에서도 overhead가 발생한다.   | Node를 순차적으로 이동해야하기 때문에 접근,탐색이 어렵고 같은 이유로 중간or맨 뒤에 데이터를 추가할때 시간이 걸린다   |
| 시간복잡도(dequeue기준) | O(1) <br>- 단 1번째 스택 값을 2번째 스택으로 옮기는 과정은 O(n)     | O(1) <br>- 단 현재위치가 아닌 곳에 값을 추가할때 O(n)   |
| 공간복잡도 | O(n)     | O(n)     |

결론적으로 현재 계산기 프로젝트에 맞는 것은 `Single LinkedList Queue`라고 생각합니다. 꼬리를 가리키는 tail을 추가하면 추가와 삭제 모두 O(1)의 시간복잡도를 가지기 때문입니다. 하지만 더블스택 큐는 새로 스택을 만드는 시점에서 O(n)의 시간복잡도를 가질 수 밖에 없기 때문에 성능이  비교적 낮다고 생각되었습니다.
> 2️⃣ 파일트리, Protocol 현재 Protocol을 따로 그룹으로 파일 분리를 해주셨는데, 이렇게 정리해주신 이유가 궁금해요! 만약 Controller 영역 관련 protocol이 생긴다면 이 protocol은 어디에 위치하게 될가요?🤔
> 
프로토콜도 타입의 일종이니 분리하는게 맞다고 생각하여 분리하였지만, 위치에 관해선 생각해 본게 없습니다. controller영역 관련 프로토콜이 생긴다면 파일트리에 대한 기준이 없는 지금은 그냥 프로토콜 폴더안에 모아 놓을것같습니다. 검색시에도 따로 결과가 나오는게 없는듯 한데 권장하는 파일트리 구조나 기준이 있을까요? 아니면 온전히 팀 또는 개인의 컨벤션인가요?
> ### 요구사항
> 1️⃣ 제네릭 말씀하신대로 오직 String만 넣어준다면 문제될 게 없습니다. 하지만 `오직` Strung만 취급한다는게 단점이 될 것 같아요. 확장성과 범용성이 떨어진다고 생각합니다. 물론 해당 프로젝트에서는 redmango의 의견대로 String만을 취급해도 문제될 게 없을 수도 있겠네요! 하지만 개인적으로 제네릭을 사용해서 확장성과 범용성을 고려한 더 나은 코드를 지향하는게 좋다고 생각해요. redmango의 생각은 어떤가요?
> 
확장성과 범용성면에서 mimm의 말씀이 맞다고 생각되어 제네릭을 활용하기로 결정했습니다!
>
> 2️⃣ Unit Test test를 좀 더 세분화하고 다양한 경우와 edge case를 고려한 test가 이뤄지면 좋을 것 같아요.😀 아직 익숙하지 못해서 다양한 경우가 떠오르지 않는 것 뿐이라고 생각해요! 조금만 더 생각하면 좋을 것 같습니다!

노력해 봤습니다...ㅠㅠ

> 3️⃣ UML 작성해주신 UML이 익숙하지 못한 디자인인 것 같아요!

수정했습니다!

File: Calculator/Calculator/CalculatorItemQueue.swift [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+//
+//  Created by redmango1446 on 2023/05/30.
+//
+
+struct CalculatorItemQueue<T>: CalculateItem {
+    private var list: LinkedList<T> = LinkedList()
+    var first: T? { return list.first?.data }
+    var last: T? { return list.last?.data }
+    var count: Int { return list.count }
+    var isEmpty: Bool { return list.isEmpty }
```
Comment (JoSH0318) [inline Calculator/Calculator/CalculatorItemQueue.swift]: LinkedList의 연산프로퍼티와 컨벤션이 다릅니다.
또란, 연산 프로퍼티는 연산과정을 return하는 것이니 연산이 잘 보여지도록 개행하는게 좋아보입니다.

File: Calculator/Calculator/LinkedList.swift [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+final class LinkedList<T> {
+    private var head: Node<T>?
+    private var tail: Node<T>?
+    
+    var isEmpty: Bool {
+        if head == nil {
+            return true
+        }
+        return false
```
Comment (JoSH0318) [inline Calculator/Calculator/LinkedList.swift]: if 문에서 early exist를 하고있습니다.  때문에 guard가 더 어울린다고 생각합니다.
만약 if문을 사용하고 싶다면 if-else로 구현해야 더 옳다는 생각을해요.
이 부분에 대해서 어떻게 생각하시나요?

File: Calculator/Calculator/LinkedList.swift [hunk 8]
```diff
@@ -0,0 +1,62 @@
+//
+//  LinkedList.swift
+//  Calculator
+//
+//  Created by redmango1446 on 2023/06/01.
+//
+
+final class LinkedList<T> {
```
Comment (JoSH0318) [inline Calculator/Calculator/LinkedList.swift]: LinkedList를 class로 구현한 이유가 있을까요?

File: Calculator/Calculator/LinkedList.swift [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        return false
+    }
+    
+    var first: Node<T>? {
+        return head
+    }
+    
+    var last: Node<T>? {
+        return tail
+    }
```
Comment (JoSH0318) [inline Calculator/Calculator/LinkedList.swift]: Node 타입을 반환하고 있습니다. 
first와 last는 CalculatorItemQueue에서 언급되고, 결국 Node 타입을 받아보고 dot 문법을 통해 data에 접근하게 됩니다.
여기서 생각해봐야 한 점이 
1️⃣ CalculatorItemQueue 입장에서는 Node타입을 알게 돼요. 사실 Node에는 internal 프로퍼티가 있고, CalculatorItemQueue는 모든 프러퍼티를 알게 된 것이죠! CalculatorItemQueue 입장에서 알고 싶은건 `요구한 값 = T`일 거예요.

2️⃣ 만약 Node가 수정됐을 때, 예를 들어 data라는 프로퍼티 명이 변경된다면 LinkedList, CalculatorItemQueue 둘다 변경해줘야합니다. 

때문에 dequeue와 마찬가지로 T를 반환하는게 더 옳다는 생각을 합니다.
이 부분에 대해서 망고의 의견이 듣고싶어요.

File: Calculator/Calculator/LinkedList.swift [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            return 0
+        }
+        
+        var count = 1
+        while let next = head?.next {
+            head = next
+            count += 1
+        }
+        return count
+    }
```
Comment (JoSH0318) [inline Calculator/Calculator/LinkedList.swift]: count가 상당히 무거워졌다는 생각이 들어요.
만약 element가 n개라면 count를 언급할 때, 시간복잡도가 O(n)이 되겠네요!
제 생각에는 `private(set) var count = 0`을 선언하고 enqueue, dequeue할 때마다 count를 제어하면 될 것 같다는 생각을 합니다.

File: Calculator/Calculator/LinkedList.swift [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            tail = newData
+            return
+        }
+        
+        tail?.next = newData
+        tail = newData
+    }
+    
+    func popFirst() -> T? {
+        guard let data: T = head?.data else {
```
Comment (JoSH0318) [inline Calculator/Calculator/LinkedList.swift]: 이 부분만 타입 명시를 하셨습니다! 특별한 이유가 있을까요? 
일관된 컨벤션이 필요해보여요!

**Review by JoSH0318 — COMMENTED** [review]
### 답변
1️⃣ Q: 프로토콜도 타입의 일종이니 분리하는게 맞다고 생각하여 분리하였지만, 위치에 관해선 생각해 본게 없습니다. controller영역 관련 프로토콜이 생긴다면 파일트리에 대한 기준이 없는 지금은 그냥 프로토콜 폴더안에 모아 놓을것같습니다. 검색시에도 따로 결과가 나오는게 없는듯 한데 권장하는 파일트리 구조나 기준이 있을까요? 아니면 온전히 팀 또는 개인의 컨벤션인가요?

A: 네. 팀/개인의 컨벤션 문제입니다. 😀 위치에 대한 이유만 있다면 무엇이든 상관없습니다! 제가 의문을 제기한 점도 "저는 이런 생각인데 망고의 생각은 무엇인가"를 여쭙고 싶었던 것 같아요! 이번 기회에 이것 또한 고려해보면 좋겠습니다!

### 요구사항
1️⃣ UML
UML을 다시 한번 생각해보면 좋을 것 같아요! protocol 이름 확인해주시고 각 객체간의 관계를 좀더 생각해보시면 좋을 것 같아요!

File: Calculator/Calculator/CalculatorItemQueue.swift (hunk 6 above)
Comment (redmango1447) [reply Calculator/Calculator/CalculatorItemQueue.swift]: 수정 했습니다!

File: Calculator/Calculator/LinkedList.swift (hunk 7 above)
Comment (redmango1447) [reply Calculator/Calculator/LinkedList.swift]: 음...확실히 guard의 이미지?가 early exist에 좀 더 적합하다고 생각됩니다. 수정 하겠습니다

File: Calculator/Calculator/LinkedList.swift (hunk 10 above)
Comment (redmango1447) [reply Calculator/Calculator/LinkedList.swift]: 앗 enqueue와 dequeue만 신경쓰다보니 count는 시간복잡도를 전혀 고려하지 않고 구현했네요! 밈이 말씀하신 방법이 좀 더 적합하다고 생각되어 수정했습니다.

File: Calculator/Calculator/LinkedList.swift (hunk 11 above)
Comment (redmango1447) [reply Calculator/Calculator/LinkedList.swift]: ㅠㅠ 피로로 인한 휴먼에러로 추측됩니다. 사실 저도 저 부분만 타입 명시를 굳이 왜 했는지 모르겠군요.

File: Calculator/Calculator/LinkedList.swift (hunk 9 above)
Comment (redmango1447) [reply Calculator/Calculator/LinkedList.swift]: 사실 처음엔 큐 타입을 호출해서 사용하는 곳에서 data만 필요하니 큐타입에서 data를 반환하면 되겠지 하는 얕은 생각을 했습니다.
지금 보니 List는 Node를 아는게 맞지만 큐의 입장에선 굳이 알 필요가 없다는 생각이 듭니다. 코드 수정했습니다!

File: Calculator/Calculator/LinkedList.swift (hunk 8 above)
Comment (redmango1447) [reply Calculator/Calculator/LinkedList.swift]: LinkedList는 Node에 다음 Node의 주소값을 가지고 있고 head와 tail은 그 주소값을 가리키고(가지고?) 있다는 점을 생각해 class로 선언했습니다만 지금보니 그건 Node가 class 타입이여야 할 이유지 LinkedList가 class인 이유가 아닌것 같습니다. ㅠㅠ LinkedList가 class여야할 이유가 있나 생각해봤는데 딱히 이유는 생각나지 않습니다......으헝헝 수정하겠습니다
